tools/bin/fetch_docs.js

```diff
@@ -25,42 +25,103 @@ var child_process = require("child_process");
 var yaml          = require("js-yaml");
 
 // constants
-var DEFAULT_REPO_PATH = "README.md";
+var DEFAULT_REPO_PATH        = "README.md";
+var DEFAULT_PERMALINK_PREFIX = "/gen/";
 
-function generateFrontMatter (useDirectSrcURI, filePath, name, version) {
-    var obj                 = {};
+function generateFrontMatter(fetchedFile) {
 
-    if(useDirectSrcURI) {
-        var p               = /https:\/\/raw.githubusercontent.com\/([^\/]+)\/([^\/]+)\/([^\/]+)\/(.+)/;
-        var filePathSplit   = filePath.match(p);
-        name                = filePathSplit[1] + "/" + filePathSplit[2];
-        version             = filePathSplit[3];
-        filePath            = filePathSplit.slice(4).join("/");
+    var frontMatterConfig = {};
+
+    frontMatterConfig.edit_link = fetchedFile.editLink;
+    frontMatterConfig.permalink = fetchedFile.permalink;
+
+    // set special values for plugins
+    if (isPluginName(fetchedFile.packageName)) {
+        frontMatterConfig.plugin_name    = fetchedFile.packageName;
+        frontMatterConfig.plugin_version = fetchedFile.version;
     }
 
-    obj.edit_link           = getRepoURIForFrontMatter(name, version, filePath);
-    obj.plugin_name         = name;
-    obj.plugin_version      = version;
+    // return front matter as a string
+    var frontMatterString = "---\n" + yaml.dump(frontMatterConfig) + "---\n\n";
+    return frontMatterString;
+}
 
-    var frontMatter         = "---\n" + yaml.dump(obj) + "\n---\n\n";
-    return frontMatter;
+function isPluginName(packageName) {
+    return packageName.match(/cordova-plugin-.*/);
 }
 
-function getRepoFileURI (name, version, filePath) {
-    return "https://raw.githubusercontent.com/" + name + "/" + version + "/" + filePath;
+function getRepoFileURI(repoName, commit, filePath) {
+    return "https://raw.githubusercontent.com/" + repoName + "/" + commit + "/" + filePath;
 }
 
-function getRepoURIForFrontMatter (name, version, filePath) {
-    return "https://github.com/" + name + "/blob/" + version + "/"+ filePath
+function getRepoEditURI(repoName, commit, filePath) {
+    return "https://github.com/" + repoName + "/blob/" + commit + "/"+ filePath
 }
 
-function getLatestRelease (packageName) {
+function getLatestRelease(packageName) {
     var latestRelease = child_process.execSync("npm info " + packageName + " dist-tags.latest");
     return latestRelease.toString().trim();
 }
 
-function getPackageName (fileConfigSrc) {
-    return fileConfigSrc.packageName || fileConfigSrc.repoName.split('/')[1];
+function packageNameFromRepoName(repoName) {
+    var repoSplit      = repoName.split('/');
+    var repoOwner      = repoSplit[0];
+    var actualRepoName = repoSplit[1];
+    return actualRepoName;
+}
+
+function getFetchedFile(entry) {
+
+    // get entry components
+    var srcConfig  = entry.src;
+    var destConfig = entry.dest;
+
+    // validate entry
+    if (!srcConfig) {
+        console.error("entry '" + entry.toString() + "' missing 'src'");
```

Comment:
I opted for the "callee prints error and returns null, caller checks for null" instead of the "callee throws exception, caller catches it" pattern. Is one more advantageous than the other?